### DISCUSSION TITLE
Can't use getValues to read a field array.

### DISCUSSION THREAD

**Discussion by adriantrunzo** [react-hook-form/react-hook-form#3114] [Q&A] [answered]
HI all,

I was surprised to find that I cannot read the current state of a field array with the `getValues` method. For example, given the following field array declaration:

```
const { fields, append } = useFieldArray({ control, name: 'myarray' });
```

The documentation led me to believe that I could then call `getValues('myarray')` (eg in a callback) to get the current state of the array, especially since I can write `useWatch({ name: 'myarray', ... })` elsewhere. I expected the output of `getValues` to be the same as `useWatch`, an array of objects.

The use case I find myself in is wanting to calculate the default value of the next element of field array before it is appended, with the calculation depending on the current elements. Ideally, I'd like to do this without watching the field array and without having to know the length of the field array. For example:

```
const wrappedAppend = useCallback(data => {
  const fields = getValues('myarray')
  const defaultValue = fields.reduce(...)

  append({ defaultValue, ...data })
}, [append, getValues])
```

My current workaround is to hoist up that wrappedAppend method to the same level as `useFieldArray` and use access to the length property of the field array, like so:

```
const count = myarray.length
const wrappedAppend = useCallback(data => {
  const inputs = []

  for (let i = 0; i < count; i++) {
    const a = getValues(`myarray[${i}].a`)
    const b = ...

    inputs.push({ a, b, ... })
  }

  const defaultValue = inputs.reduce(...)

  append({ defaultValue, ...data })
}, [append, getValues])
```

Am I misunderstanding the documentation? If `useWatch` can take the name of the field array, why can't `getValues`? This is the sandbox I used to help me clarify my understanding: https://codesandbox.io/s/use-field-array-getvalues-97cd6.

Thanks!

**Comment by bluebill1049** [comment, marked as answer]
getValues is always going to be behind the render, because it's an action to read what's in the form not subscribed to the form.

**Reply by adriantrunzo** [reply]
Thanks for the reply. I understand that `getValues` is an action and is not subscribing and that is the behavior I am looking for. This is a one time calculation based on the current inputs of the field array before appending the next one. I am not sure I understand why it can't be used to read the current state of all of the inputs in the field array. If it can read a specific input in the array, it seems like it could also read all of the current inputs in the field array.

**Reply by bluebill1049** [reply]
oh you try to read the `key`, i would use `getValues()` we don't support the key look up

**Reply by adriantrunzo** [reply]
For example, in the sandbox, you can't `console.dir(getValues('myarray'))` to just print the current state of the array. You can do a one time read of a specific input `console.dir(getValues('myarray[0].something'))`.

Maybe what I am looking for is a feature request? That is, if you call `getValues` with just the name of field array and not a specific index/input, it returns the current state of all the inputs registered in all of the fields of the array.

**Reply by adriantrunzo** [reply]
Here is a better explanation of my expectation:

Given the field array 'myarray', which currently has X entries, each with 2 registerted inputs 'a' and 'b', I expect:

```
const current = getValues('myarray')
-> [
  0: { a: ..., b: ... },
  1: { a: ..., b: ... },
  2: { a: ..., b: ... }
  ...
  X-1: { a: ..., b: ... }
]
```

Which would be equivalent to me looping through the field array via index:

```
const current = []

 for (let i = 0; i < X; i++) {
  current.push({ a: getValues(`myarray[${i}].a`), b: getValues(`myarray[${i}].b`) })
}
```

Can `getValues('myarray')` just do that loop for me?

**Reply by adriantrunzo** [reply]
Thanks again for the response. Maybe the general case isn't useful and I have a work around.

**Reply by bluebill1049** [reply]
`Can getValues('myarray') just do that loop for me?`

It's under the consideration right now. but I think it's an overall API update, then the same API needs to be applied for `trigger` as well. will keep you updated if i get to that phase of updating all API to support key look up.

**Reply by bluebill1049** [reply]
for now, you probably have to use `getValues().keyName`

**Reply by adriantrunzo** [reply]
Ah, now I understand what you mean by ‘key’! Thanks for the further clarification and sorry for my confusion. The API design makes sense. 

I didn’t think of ‘getValues().keyName’ so I will keep that in mind for future.

**Reply by bluebill1049** [reply]
👍

**Reply by jgreens** [reply]
@bluebill1049 @adriantrunzo, I know this is an old thread – just leaving a question in case you get a chance to see it. I think my question is related as using `getValues().keyName` fixed the issue I am facing, but I am not sure why.

In this [example from the docs](https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-m8w6j?file=/src/fieldArray.js:1012-1043), in line `45` of `fieldArray.js`, can you explain why `setValue` with `getValues().key` must be used instead of `prepend` from the `useFieldArray` API? When I was using `prepend` – my components didn't seem to be updating/rerendering properly.

Thank you.

**Reply by bluebill1049** [reply]
@jgreens which version are you using? can you provide a codesandbox?

**Reply by jgreens** [reply]
Hi – replied on a different thread that seemed to match my question better:
https://github.com/react-hook-form/react-hook-form/discussions/3702#discussioncomment-1416171

I am using the v7.11.1, and am just asking about the example in your codesandbox:
https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-m8w6j?file=/src/fieldArray.js:1043-1086
